File: Array/001-TwoSum/TwoSum.js
//Brute force way (Javascript) :

var twoSum = function(nums, target) {
    for (let idx1 = 0; idx1 < nums.length; idx1++) {
        for (let idx2 = idx1+1; idx2 < nums.length; idx2++) {
            if (nums[idx1] + nums[idx2] == target) {
                return [idx1,idx2]
            }
        }
    }
};

//Optimal way using hasmap (Javascript) :

var twoSum = function(nums, target) {
    const numMap = new Map()

    for (idx = 0; idx < nums.length; idx++) {
        const remainder = target - nums[idx]
        if (numMap.has(remainder)) {
            return [numMap.get(remainder), idx]
        }
        numMap.set(nums[idx], idx)
    }
};

File: Array/001-TwoSum/TwoSum.go
// Brute force way 

func twoSum(nums []int, target int) []int {
    for idx1, num1 := range nums {
        for idx2, num2 := range nums {
            if (num1+num2 == target) && (idx1 != idx2) {
                return []int{idx1,idx2}
            }
        }
    }
    return []int{}
}

// Optimal way (using hash map)

func twoSum(nums []int, target int) []int {
    hashMap := make(map[int]int)

    for idx, num := range nums {
        remainder := target - num
        idx2, ok := hashMap[remainder]
        if ok {
            return []int{idx2, idx}
        }
        hashMap[num] = idx
    }

    return []int{}
}
